### ISSUE TITLE
Shipping taxes incorrectly calculated due to rounding change from #17563

### ISSUE THREAD

**Issue by hjoelr** [woocommerce/woocommerce#17779]
It seems that the one cent off bug that I fixed in #17563 has caused another bug in shipping tax calculations. I wish I would have caught this before 3.2.4 was released....

Anyway, the tax calculations in cart and on checkout can be slightly different from what is displayed on the order details screen. After spending a significant amount of time trying to figure why, it's pointed to the the change that added the `intval` and `round` functions here.

https://github.com/woocommerce/woocommerce/blob/36e7bb04a158cd6e4c286809d870436341a664ea/includes/wc-core-functions.php#L1500-L1503

I believe that causes an issue on line 352 below:

https://github.com/woocommerce/woocommerce/blob/36e7bb04a158cd6e4c286809d870436341a664ea/includes/class-wc-cart-totals.php#L346-L360

In the scenario I encountered, this was the value of `$shipping_object->taxes` when this was called:

```php
$shipping_object->taxes = array(
	2 =>  .0864,
	92 => .0364,
);
```

After the call to `wc_add_number_precision_deep()`, `$shipping_line->taxes` equals the following:

```php
$shipping_line->taxes = array(
	2 =>  9,
	92 => 4,
);
```

I believe that the shipping calculation code assumes that the returned value would not have chopped off decimal places. This ultimately gets saved to two `shipping_tax_amount` meta fields (one for each tax line) in `order_itemmeta` with the values `0.09` and `0.04`.

![image](https://user-images.githubusercontent.com/221971/32971634-1187d566-cbbc-11e7-99b9-b5ac6591b82a.png)

As you'll notice, the `tax_amount` does not get the extra decimal places removed whereas the `shipping_tax_amount` does.

Also, the `_order_shipping_tax` gets calculated to $0.12...

![image](https://user-images.githubusercontent.com/221971/32972020-0a38036a-cbbe-11e7-8b70-fa088020ece8.png)

...but tax values of `0.09` and `0.04` add to $0.13.

I'm not really sure what to do to fix this. If we remove the `intval` and `round` functions in `wc_add_number_precision()`, that will cause the issue where some discounts were one cent less than they should be (the problem described in #17563).

All said, I don't exactly know what's happening, but I think it's related to #17663.

## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [ ] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Have two levels of tax set up with shipping selected similar to this:

![image](https://user-images.githubusercontent.com/221971/32970102-921816f2-cbb5-11e7-8b2a-f02392c431bb.png)

2. Add an item to the cart that is $13.95.
3. Checkout with a user whose address matches both tax levels.
3. Notice the total tax calculation in the cart and checkout is $1.07.
4. Place the order and see on the order confirmation screen the tax is $1.06.

## Expected/actual behavior

When I removed the `intval` and `round` code inside `wc_add_number_precision()`, the in-cart and on checkout tax was $1.06 and so was the order details tax amount.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

**Comment by mikejolley** [comment]
@hjoelr Need some more details. What was the shipping cost? Tax settings? I wasn't able to replicate with the above instructions.

Alternatively, a failing unit test would be helpful.

**Comment by mikejolley** [comment]
@hjoelr Can you update me on this so I can work on it today?

**Comment by ChromeOrange** [comment]
@mikejolley I might be seeing the same issue

![](https://cl.ly/2H0D0t0B3Y40/Image%202017-11-20%20at%2012.40.44%20pm.png)

Tax settings : (changing the rounding option makes no difference)

![](https://cl.ly/092c3u3g1642/Image%202017-11-20%20at%2012.43.48%20pm.png)

**Comment by hjoelr** [comment]
@mikejolley Sorry for not replying sooner. I didn't see your comment until this morning and I wouldn't have had a chance to get to it anyway until now. In my case the shipping was $1.82. Here's my tax settings:

![image](https://user-images.githubusercontent.com/221971/33020046-bafd732a-cdca-11e7-8605-a1cdd82ea376.png)

It looks like maybe @ChromeOrange was able to replicate it. I was hoping to be able to submit a unit test for this, but ran out of time on Friday. It also wasn't clear how to go about testing it. Let me know if you still need me to get you more information.

**Comment by mikejolley** [comment]
The order view is rounded per line so I don't think thats related but I'll see if we can write a test case for this one once I am able to.

**Comment by hjoelr** [comment]
@mikejolley You may check testing `WC_Order->get_taxes()` array against `WC_Order->get_total_tax()`

**Comment by AKSUTAH** [comment]
I just experienced a similar issue that is pretty serious.
An order was placed and the order confirmation was sent to the customer for the amount of $120.17.
The order was authorized for that amount, $120.17...
When I tried to capture, it was rejected because the amount I was trying to capture was more than authorized.
I looked at the order and noticed that the total was now $120.18...
No changes had been made to the order. Upon further inspection, I see that a state tax item of 3.808 was rounded to the amount of 3.81.
How this occurred after the fact is the real problem.
